src/target/riscv/riscv-013.c

```diff
@@ -4274,62 +4373,57 @@ static int read_memory(struct target *target, target_addr_t address,
 	if (count == 0)
 		return ERROR_OK;
 
-	if (size != 1 && size != 2 && size != 4 && size != 8 && size != 16) {
+	if (!IS_PWR_OF_2(size) || size > 16) {
 		LOG_TARGET_ERROR(target, "BUG: Unsupported size for memory read: %d", size);
 		return ERROR_FAIL;
 	}
 
-	int ret = ERROR_FAIL;
-	RISCV_INFO(r);
-	RISCV013_INFO(info);
-
-	char *progbuf_result = "disabled";
-	char *sysbus_result = "disabled";
-	char *abstract_result = "disabled";
-
-	for (unsigned int i = 0; i < RISCV_NUM_MEM_ACCESS_METHODS; i++) {
-		int method = r->mem_access_methods[i];
-
-		if (method == RISCV_MEM_ACCESS_PROGBUF) {
-			if (mem_should_skip_progbuf(target, address, size, true, &progbuf_result))
-				continue;
-
-			ret = read_memory_progbuf(target, address, size, count, buffer, increment);
-
-			if (ret != ERROR_OK)
-				progbuf_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_SYSBUS) {
-			if (mem_should_skip_sysbus(target, address, size, increment, true, &sysbus_result))
-				continue;
-
-			if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 0)
-				ret = read_memory_bus_v0(target, address, size, count, buffer, increment);
-			else if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 1)
-				ret = read_memory_bus_v1(target, address, size, count, buffer, increment);
-
-			if (ret != ERROR_OK)
-				sysbus_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_ABSTRACT) {
-			if (mem_should_skip_abstract(target, address, size, increment, true, &abstract_result))
-				continue;
-
-			ret = read_memory_abstract(target, address, size, count, buffer, increment);
+	mem_access_status_t skip_reason[] = {
+		[RISCV_MEM_ACCESS_PROGBUF] = MEM_ACCESS_STATUS_DISABLED,
+		[RISCV_MEM_ACCESS_SYSBUS] = MEM_ACCESS_STATUS_DISABLED,
+		[RISCV_MEM_ACCESS_ABSTRACT] = MEM_ACCESS_STATUS_DISABLED,
+	};
 
-			if (ret != ERROR_OK)
-				abstract_result = "failed";
-		} else if (method == RISCV_MEM_ACCESS_UNSPECIFIED)
-			/* No further mem access method to try. */
-			break;
+	RISCV_INFO(r);
+	for (unsigned int i = 0; i < r->mem_access_method_num; ++i) {
+		riscv_mem_access_method_t method = r->mem_access_methods[i];
+		switch (method) {
+			case RISCV_MEM_ACCESS_PROGBUF:
+				skip_reason[method] =
+					read_memory_progbuf(target, address,
+						size, count, buffer, increment);
+				break;
+			case RISCV_MEM_ACCESS_SYSBUS:
+				skip_reason[method] =
+					read_memory_sysbus(target, address,
+						size, count, buffer, increment);
+				break;
+			case RISCV_MEM_ACCESS_ABSTRACT:
+				skip_reason[method] =
+					read_memory_abstract(target, address,
+						size, count, buffer, increment);
+				break;
+			default:
+				LOG_TARGET_ERROR(target, "Unknown memory access method: %d", method);
```

Comment:
An assert can be used here:

```suggestion
				LOG_TARGET_ERROR(target, "Unknown memory access method: %d", method);
				assert(false);
```

Reply:
Addressed